### PR TITLE
Added state for a job

### PR DESCRIPTION
**Pull request by eldario**
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes

## Description

> Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I wrote unit tests for my code
- [x] I have made changes in [CHANGELOG.md](https://github.com/avto-dev/amqp-rabbit-laravel-queue/blob/master/CHANGELOG.md) file

> About your changes in `CHANGELOG.md`:
>
> * Add new version header like `## v1.x.x`, if it does not exists
> * Add description under `added`/`changed`/`fixed` sections
> * Add reference to closed issues `[#000]`
> * Add link to issue in the end of document

### CODE REVIEW

File: tests/Feature/QueueWorkerTest.php [hunk 1]
```diff
@@ -316,4 +317,30 @@ public function testWithWorkerTimeoutAndResume(): void
...
+     *
+     * @return void
+     *
+     * @group foo
+     */
+    public function testStoringStateJob(): void
+    {
+        $this->assertFalse(Sharer::has(QueueJobWithSavedState::class . '-handled'));
+
+        $this->dispatcher->dispatch((new QueueJobWithSavedState)->delay(1)); // Should be grater then queue timeout
```
Comment (tarampampam) [inline tests/Feature/QueueWorkerTest.php]: del `->delay(1)`

File: tests/Stubs/QueueJobWithSavedState.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+use AvtoDev\AmqpRabbitLaravelQueue\Tests\Sharer\Sharer;
+
+class QueueJobWithSavedState extends SimpleQueueJob implements StoreStateInterface
+{
+    use WithJobStateTrait, InteractsWithQueue;
+
+    /**
+     * Magic name of state name.
+     */
+    public const MAGIC_PROPERTY = 'magic_counter';
```
Comment (tarampampam) [inline tests/Stubs/QueueJobWithSavedState.php]: del, change to hardcoded like `-trigered`

File: src/Job.php [hunk 3]
```diff
@@ -85,6 +90,26 @@ public function delete(): void
         $this->consumer->acknowledge($this->message);
     }
 
+    /**
```
Comment (tarampampam) [inline src/Job.php:104]: Annotation

Comment (tarampampam) [reply src/Job.php:104]: add test for exception when passed something like `function (): void {}` or `\tmpfile()`

File: src/Job.php [hunk 4]
```diff
@@ -85,6 +90,26 @@ public function delete(): void
...
+     * @param mixed $data
+     */
+    public function setMessageContext($data): void
+    {
+        $this->message->setProperty(static::CONTEXT_PROPERTY, \serialize($data));
+    }
+
+    /**
+     * @return mixed|null
```
Comment (tarampampam) [inline src/Job.php]: Annotation

File: src/Job.php [hunk 5]
```diff
@@ -85,6 +90,26 @@ public function delete(): void
...
+     */
+    public function getMessageContext()
```
Comment (tarampampam) [inline src/Job.php]: add support for some default value, like `$default = null`

Comment (tarampampam) [reply src/Job.php]: + check does vendor has method like `hasProperty`

File: src/StoreStateInterface.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace AvtoDev\AmqpRabbitLaravelQueue;
+
+/**
+ * Interface StoreStateInterface.
```
Comment (tarampampam) [inline src/StoreStateInterface.php]: delete this

File: src/StoreStateInterface.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Store state in job's message.
+ */
+interface StoreStateInterface
+{
+    /**
+     * Store state in job.
```
Comment (tarampampam) [inline src/StoreStateInterface.php]: add note about value must allows serialization

File: src/WithJobStateTrait.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace AvtoDev\AmqpRabbitLaravelQueue;
+
+use AvtoDev\AmqpRabbitLaravelQueue\Job as RabbitJob;
+
+/**
+ * @property RabbitJob $job
```
Comment (tarampampam) [inline src/WithJobStateTrait.php]: del

File: src/WithJobStateTrait.php [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+ */
+trait WithJobStateTrait
+{
+    /**
+     * {@inheritdoc}
```
Comment (tarampampam) [inline src/WithJobStateTrait.php]: copy paste annotation

File: tests/Feature/QueueWorkerTest.php [hunk 10]
```diff
@@ -316,4 +317,30 @@ public function testWithWorkerTimeoutAndResume(): void
...
+
+        $process_info = $this->startArtisan('queue:work', [], 2.0, [
+            'QUEUE_TIMEOUT' => 800,
```
Comment (tarampampam) [inline tests/Feature/QueueWorkerTest.php]: del

File: tests/Feature/QueueWorkerTest.php [hunk 11]
```diff
@@ -316,4 +317,30 @@ public function testWithWorkerTimeoutAndResume(): void
...
+            'QUEUE_RESUME'  => true,
+        ]);
+        $this->assertTrue($process_info['timed_out']);
+        $this->assertSame(QueueJobWithSavedState::COUNT_OF_TRIES,
+            Sharer::get(QueueJobWithSavedState::class . '-handled'));
+        $this->assertTrue(Sharer::has(QueueJobWithSavedState::class . '-failed'));
+        $expected_value = QueueJobWithSavedState::COUNT_OF_TRIES * QueueJobWithSavedState::ITERATION_INCREMENT;
+
+        self::assertEquals($expected_value, Sharer::get(QueueJobWithSavedState::MAGIC_PROPERTY));
```
Comment (tarampampam) [inline tests/Feature/QueueWorkerTest.php]: `self::` -> `$this->`

File: tests/JobTest.php [hunk 12]
```diff
@@ -258,5 +258,15 @@ public function testGetters(): void
...
 
         $this->assertSame($this->message, $this->job->getMessage());
+
+        // Message Context
+
+        $this->assertNull($this->message->getProperty($this->job::CONTEXT_PROPERTY));
+
+        $this->job->setMessageContext($data = ['foo_key' => 'bar_value']);
+
+        $this->assertSame(\serialize($data), $this->message->getProperty($this->job::CONTEXT_PROPERTY));
```
Comment (tarampampam) [inline tests/JobTest.php]: del

**Review by tarampampam — CHANGES_REQUESTED** [review]

File: tests/Stubs/QueueJobWithSavedState.php (hunk 2 above)
Comment (eldario) [reply tests/Stubs/QueueJobWithSavedState.php]: `changed`

File: src/StoreStateInterface.php (hunk 6 above)
Comment (eldario) [reply src/StoreStateInterface.php]: deleted

File: src/StoreStateInterface.php (hunk 7 above)
Comment (eldario) [reply src/StoreStateInterface.php]: Added

File: src/StoreStateInterface.php [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+     * @param mixed $data
+     */
+    public function setState($data);
+
+    /**
+     * Returns stored state of job.
+     *
+     * @return mixed|null
+     */
+    public function getState();
```
Comment (eldario) [inline src/StoreStateInterface.php]: Added `$default`

File: src/WithJobStateTrait.php (hunk 8 above)
Comment (eldario) [reply src/WithJobStateTrait.php]: deleted

File: src/WithJobStateTrait.php (hunk 9 above)
Comment (eldario) [reply src/WithJobStateTrait.php]: copied. pasted.

File: tests/Feature/QueueWorkerTest.php (hunk 1 above)
Comment (eldario) [reply tests/Feature/QueueWorkerTest.php]: deleted

File: tests/Feature/QueueWorkerTest.php (hunk 10 above)
Comment (eldario) [reply tests/Feature/QueueWorkerTest.php]: deleted

File: tests/Feature/QueueWorkerTest.php (hunk 11 above)
Comment (eldario) [reply tests/Feature/QueueWorkerTest.php]: changed

File: tests/JobTest.php (hunk 12 above)
Comment (eldario) [reply tests/JobTest.php]: deleted

File: src/Job.php (hunk 5 above)
Comment (eldario) [reply src/Job.php]: no `hasProperty`, `$default` added

File: src/Job.php (hunk 4 above)
Comment (eldario) [reply src/Job.php]: added

File: src/Job.php (hunk 3 above)
Comment (eldario) [reply src/Job.php:104]: Tests added

File: src/Job.php [hunk 14]
```diff
@@ -85,6 +90,35 @@ public function delete(): void
...
+     * Set the message context under a special key.
+     *
+     * @param mixed $data
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function setMessageContext($data): void
+    {
+        if (\is_resource($data)) {
+            throw new \InvalidArgumentException;
```
Comment (tarampampam) [inline src/Job.php]: `use InvalidArgumentException` once

File: src/Job.php [hunk 15]
```diff
@@ -85,6 +90,35 @@ public function delete(): void
...
+        }
```
Comment (tarampampam) [inline src/Job.php]: One empty line after `}`

File: src/Job.php [hunk 16]
```diff
@@ -85,6 +90,35 @@ public function delete(): void
...
+    /**
+     * Get the message context under a special key.
+     *
+     * @param $default
+     *
+     * @return mixed|null
+     */
+    public function getMessageContext($default = null)
+    {
+        return ($data = $this->message->getProperty(static::CONTEXT_PROPERTY, $default)) !== null
```
Comment (tarampampam) [inline src/Job.php]: `$data = $this->message->getProperty(static::CONTEXT_PROPERTY, $default)` before `return`

File: CHANGELOG.md [hunk 17]
```diff
@@ -4,6 +4,12 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog][keepachangelog] and this project adheres to [Semantic Versioning][semver].
 
+## Unreleased
+
+### Added
+
+- Set the message context under a special key for a `job`
```
Comment (tarampampam) [inline CHANGELOG.md]: - Added feature for storing `job` state between job tries

**Review by tarampampam — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 17 above)
Comment (eldario) [reply CHANGELOG.md]: `Modified`

File: src/Job.php [hunk 18]
```diff
@@ -65,6 +75,10 @@ public function __construct(Container $container,
         $this->message          = $message;
         $this->connectionName   = $connection_name;
         $this->delayed_exchange = $delayed_exchange;
+
+        $current_state = $message->getProperty(static::CONTEXT_PROPERTY, 'N;');
```
Comment (tarampampam) [inline src/Job.php]: `'N;'` is evil

File: src/Job.php [hunk 19]
```diff
@@ -65,6 +75,10 @@ public function __construct(Container $container,
...
+
+        $this->state = \unserialize($current_state) ?? new JobState;
```
Comment (tarampampam) [inline src/Job.php]:
```php
$this->state = $current_state !== null
    ? \unserialize($current_state)
    : new JobState;
```

File: src/JobState.php [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+use Closure;
+use InvalidArgumentException;
+use Illuminate\Support\Collection;
+
+class JobState extends Collection implements JobStateInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function serialize()
```
Comment (tarampampam) [inline src/JobState.php:16]: add test

**Review by tarampampam — CHANGES_REQUESTED** [review]

File: src/JobState.php (hunk 20 above)
Comment (eldario) [reply src/JobState.php:16]: `added`

File: src/Job.php (hunk 19 above)
Comment (eldario) [reply src/Job.php]: `fixed`

File: src/Job.php (hunk 18 above)
Comment (eldario) [reply src/Job.php]: > No more evil

File: src/Job.php (hunk 16 above)
Comment (eldario) [reply src/Job.php]: > `removed` method

File: src/Job.php (hunk 15 above)
Comment (eldario) [reply src/Job.php]: `removed`

File: src/Job.php (hunk 14 above)
Comment (eldario) [reply src/Job.php]: `removed`

File: src/StoreStateInterface.php (hunk 13 above)
Comment (eldario) [reply src/StoreStateInterface.php]: `removed`

File: CHANGELOG.md [hunk 21]
```diff
@@ -4,6 +4,12 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog][keepachangelog] and this project adheres to [Semantic Versioning][semver].
 
+## Unreleased
+
+### Added
+
+- Added feature for storing `job` state between `job` tries
```
Comment (tarampampam) [inline CHANGELOG.md]: Double `Added` - remove

File: CHANGELOG.md [hunk 22]
```diff
@@ -4,6 +4,12 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog][keepachangelog] and this project adheres to [Semantic Versioning][semver].
 
+## Unreleased
```
Comment (tarampampam) [inline CHANGELOG.md]: `## v2.1.0`

File: README.md [hunk 23]
```diff
@@ -181,7 +181,12 @@ You can dispatch your jobs as usual (`dispatch(new Job)` or `dispatch(new Job)->
 
 - Jobs delaying (plugin `rabbitmq_delayed_message_exchange` for RabbitMQ server is required);
 - Jobs priority (job should implements `PrioritizedJobInterface` interface);
-- Automatically delayed messages exchanges bindings (only if you use command `rabbit:setup` for queues and exchanges creation).
+- Automatically delayed messages exchanges bindings (only if you use command `rabbit:setup` for queues and exchanges creation);
+- The ability to store the state of `job`
+
+#### Store state
```
Comment (tarampampam) [inline README.md]: `State storing`

File: src/Job.php [hunk 24]
```diff
@@ -22,6 +22,11 @@ class Job extends \Illuminate\Queue\Jobs\Job implements JobContract
      */
     public const ATTEMPTS_PROPERTY = 'job-attempts';
 
+    /**
+     * Store state property name.
+     */
+    public const CONTEXT_PROPERTY = 'job-context';
```
Comment (tarampampam) [inline src/Job.php]: to `job-state`

Comment (tarampampam) [reply src/Job.php]: `CONTEXT_PROPERTY` -> `STATE_PROPERTY`

File: src/JobState.php [hunk 25]
```diff
@@ -0,0 +1,40 @@
+<?php
+
```
Comment (tarampampam) [inline src/JobState.php:2]: `declare(strict_types = 1);` into each class

File: src/JobState.php [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+namespace AvtoDev\AmqpRabbitLaravelQueue;
+
+use Closure;
+use InvalidArgumentException;
+use Illuminate\Support\Collection;
+
+class JobState extends Collection implements JobStateInterface
```
Comment (tarampampam) [inline src/JobState.php:11]: Where is `JobStateTest`?

File: src/WithJobStateTrait.php [hunk 27]
```diff
@@ -0,0 +1,31 @@
...
+namespace AvtoDev\AmqpRabbitLaravelQueue;
+
+use RuntimeException;
+
+/**
+ * Trait WithJobStateTrait.
+ *
+ * @property Job $job
+ */
+trait WithJobStateTrait
```
Comment (tarampampam) [inline src/WithJobStateTrait.php:12]: Where is `WithJobStateTraitTest`?

File: src/WithJobStateTrait.php [hunk 28]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace AvtoDev\AmqpRabbitLaravelQueue;
+
+use RuntimeException;
+
+/**
+ * Trait WithJobStateTrait.
```
Comment (tarampampam) [inline src/WithJobStateTrait.php]: Delete unimportant annotations from each file/method

**Review by tarampampam — CHANGES_REQUESTED** [review]

File: src/WithJobStateTrait.php (hunk 28 above)
Comment (eldario) [reply src/WithJobStateTrait.php]: `deleted`

File: src/JobState.php (hunk 26 above)
Comment (eldario) [reply src/JobState.php:11]: `added`

File: src/JobState.php (hunk 25 above)
Comment (eldario) [reply src/JobState.php:2]: `added`

File: src/Job.php (hunk 24 above)
Comment (eldario) [reply src/Job.php]: `changed`

File: README.md (hunk 23 above)
Comment (eldario) [reply README.md]: `changed`

File: CHANGELOG.md (hunk 22 above)
Comment (eldario) [reply CHANGELOG.md]: `changed`

File: CHANGELOG.md (hunk 21 above)
Comment (eldario) [reply CHANGELOG.md]: `removed`

File: src/WithJobStateTrait.php (hunk 27 above)
Comment (eldario) [reply src/WithJobStateTrait.php:12]: `added`

File: tests/JobStateTest.php [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+     * @return void
+     */
+    public function testUnserializeMethod(): void
+    {
+        $data = \serialize($items = ['some_items' => ['foo', 'bar', 23]]);
+        $this->job_state_class->unserialize($data);
+
+        $this->assertSame($items, $this->job_state_class->all());
+    }
+
```
Comment (tarampampam) [inline tests/JobStateTest.php:68]: I wanna next test:

```php
$data = [123, 'foo' => 'bar', false, null, \M_PI];

$this->assertSame($data, \unserialize(\serialize(new JobState($data)))->all());
```

**Review by tarampampam — CHANGES_REQUESTED** [review]
One more test is required

File: tests/JobStateTest.php (hunk 29 above)
Comment (eldario) [reply tests/JobStateTest.php:68]: `added`

**Review by tarampampam — APPROVED** [review]

**Comment by tarampampam** [comment]
@eldario Great job! Thx! 👍

**Pull request merged**